File: packages/signals/src/prepare.ts
import type { Review } from '@signalyze/shared';
import { dayNumber, monthKey } from './math';
import {
  charNgrams,
  codePointLength,
  detectLanguage,
  normalizeText,
  tokenize,
  type EngineLanguage,
} from './text';

/** A review plus the derived fields every signal needs. Computed once. */
export interface PreparedReview {
  review: Review;
  day: number;
  month: string;
  normalizedText: string;
  textLength: number;
  tokens: string[];
  language: EngineLanguage;
  normalizedOwnerResponse: string | null;
}

export function prepareReviews(reviews: readonly Review[]): PreparedReview[] {
  return reviews.map((review) => {
    const normalizedText = normalizeText(review.text);
    const tokens = tokenize(normalizedText);
    return {
      review,
      day: dayNumber(review.date),
      month: monthKey(review.date),
      normalizedText,
      textLength: codePointLength(review.text.trim()),
      tokens,
      language: detectLanguage(tokens),
      normalizedOwnerResponse:
        review.ownerResponse !== null && review.ownerResponse.trim() !== ''
          ? normalizeText(review.ownerResponse)
          : null,
    };
  });
}

export function ngramsOf(prepared: PreparedReview, n: number): Set<string> {
  return charNgrams(prepared.normalizedText, n);
}

File: apps/api/signalyze_api/engine/prepare.py
"""Derived per-review fields, computed once. Twin of ``packages/signals/src/prepare.ts``."""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass

from signalyze_api.engine.mathutil import day_number, month_key
from signalyze_api.engine.text import (
    EngineLanguage,
    char_ngrams,
    code_point_length,
    detect_language,
    js_trim,
    normalize_text,
    tokenize,
)
from signalyze_api.models import Review


@dataclass(slots=True)
class PreparedReview:
    """A review plus the derived fields every signal needs."""

    review: Review
    day: int
    month: str
    normalized_text: str
    text_length: int
    tokens: list[str]
    language: EngineLanguage
    normalized_owner_response: str | None


def prepare_reviews(reviews: Sequence[Review], ngram_size: int) -> list[PreparedReview]:
    del ngram_size  # kept for signature parity with the reference
    prepared: list[PreparedReview] = []
    for review in reviews:
        normalized_text = normalize_text(review.text)
        tokens = tokenize(normalized_text)
        owner = review.ownerResponse
        prepared.append(
            PreparedReview(
                review=review,
                day=day_number(review.date),
                month=month_key(review.date),
                normalized_text=normalized_text,
                text_length=code_point_length(js_trim(review.text)),
                tokens=tokens,
                language=detect_language(tokens),
                normalized_owner_response=(
                    normalize_text(owner) if owner is not None and js_trim(owner) != "" else None
                ),
            )
        )
    return prepared


def ngrams_of(prepared: PreparedReview, n: int) -> set[str]:
    return char_ngrams(prepared.normalized_text, n)
